### PR TITLE
Fixed per comments on pull request #831. Replacement for pull request #831.  Now includes ExternProto expander

### PR DESCRIPTION
**Pull request by coderextreme**

### CODE REVIEW

**Comment by coderextreme** [comment]
It’s in src/util/json/PrototypeExpander.js (grab most recent).  I test for $ before calling $.ajaxSetup or $.get.  I have a need for the ExternPrototypeExpander to work on clients in X3DOM, but on servers in X3DJSONLD (for testing and debugging purposes), so we can get rid of the Extern code, but that means X3DOM is less functional without a server, and I have to maintain 2 code bases (which I do anyway—3 for JSONParser—trying to make things simpler).  I *just* integrated the Extern expander, and would like to keep it in my code.   Thus I need $.get or something.

I’ve got a pure get somewhere that uses promises, but I still need to use $.ajaxSetup to set async false.  Got a fix for that?

So a .get that works on both client and server, that works with relative and absolute URLs would be most desirable.

As far as I can see x3dom and JQuery appear to be somewhat compatible, since I use them together a lot. Why do you see it as a blocker?  I agree that we should not depend on JQuery and I will look for another answer.  I have a get in my HTML code in X3DJSONLD.   I also use $ in X3DJSONLD.js when I shouldn’t (it’s not designed to be dependent on JQuery), I just would like something open source and easy in my base libraries for X3DJSONLD that I don’t have to maintain or really think about.  However, I may want to start using promises, and I don’t think $.get works too well with promises.  Basically, I want something that will load a file from the file system or the web server without much fuss.  If there was something to replace loadURLs and processURLs I would use it if I knew what it was.

Is there a good example of using x3dom requestmanager?  Perhaps I can start using X3DOM more integrated into X3DJSONLD, instead of as an external dependency (with my own build).

I probably also need to integrate the DOM2JSONSerializer for XML extern protos.   More work.

John

Sent from Mail for Windows 10

**Comment by andreasplesch** [comment]
#688 introduces RequestManager. It is used for Inline, and applications.

You can set async for XHR:

https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/Synchronous_and_Asynchronous_Requests

Could you run a (express or python) server on the server to serve files to local x3dom on the server ?

x3dom is already too heavy and should not require more dependencies.

**Comment by coderextreme** [comment]
Yes, but that doesn’t work on my webhost which only supports PHP for the most part. Will have to write something in PHP eventually.  I like client side and I got it to work with sync XHR.

John

Sent from Mail for Windows 10

**Comment by coderextreme** [comment]
I have removed the dependency on jquery, even though I never planned to require that all X3DOM users use it.  The code is much better, and it’s pretty easy to do synchronous xhr.

John

Sent from Mail for Windows 10

File: src/nodes/Core/X3DNode.js [hunk 1]
```diff
@@ -293,6 +293,9 @@ x3dom.registerNodeType(
                                 case "string":
                                     this._vf[field] = msg;
                                     break;
+				default:
```
Comment (andreasplesch) [inline src/nodes/Core/X3DNode.js]: not sure about this default. How is it necessary ?

File: src/util/json/PrototypeExpander.js [hunk 2]
```diff
@@ -0,0 +1,1229 @@
...
+				
+		}
+		// console.error("Processed URLs", localArray.join(" "));
+		return localArray;
+	},
+
+	/**
+	 * Use almost any method possible to load a set of URLs.  The loadpath is the
+	 * original URL the main JSON got laoded from, Urls is the se of urls, and
+	 * the loadedCallback returns the data and the URL it was loaded from.
```
Comment (andreasplesch) [inline src/util/json/PrototypeExpander.js]: Perhaps add that the http(s) and fs modules are required on server with node.js ?

**Comment by andreasplesch** [comment]
Here are test builds with the latest commits merged:

http://x3dom-builds.surge.sh/build_5/
http://x3dom-builds.surge.sh/build_5/x3dom-full.debug.js

In fact, each time commits are added to this PR, a new build is triggered automatically and pushed to http://x3dom-builds.surge.sh

**Comment by coderextreme** [comment]
I removed the default case and added comments for loadURLs on the server

**Comment by andreasplesch** [comment]
Here is a test page to load json inlines:
https://x3dom-json-inline-test.glitch.me/index.xhtml

**Comment by andreasplesch** [comment]
I had to revert since DataUrls are not correctly recognized, eg. will be assumed .json. They could be either .x3d or .json, once decoded . The mime type for both would be text/plain . Not sure. May be check how x_ite distinguishes between .x3d and .wrl for DataUrls.

**Comment by coderextreme** [comment]
I believe x_ite distinguishes by trying to parse and failing.

John.

File: src/nodes/Networking/Inline.js [hunk 3]
```diff
@@ -159,9 +159,23 @@ x3dom.registerNodeType(
...
-                    xhr.overrideMimeType('text/xml');   //application/xhtml+xml
+                if (this._vf.url.length && this._vf.url[0].length) {
+			if (this._vf.url[0].toLowerCase().endsWith(".x3d")) {
+				isJSON = false;
+				if (xhr.overrideMimeType)
+				    xhr.overrideMimeType('text/xml');   //application/xhtml+xml
+			} else if (this._vf.url[0].toLowerCase().endsWith(".json")) {
+				if (xhr.overrideMimeType)
+				    xhr.overrideMimeType('application/json');
+			}
```
Comment (andreasplesch) [inline src/nodes/Networking/Inline.js]: handle other cases like DataURL

File: src/nodes/Networking/Inline.js [hunk 4]
```diff
@@ -159,9 +159,23 @@ x3dom.registerNodeType(
               
                 var that = this;
 
+		var isJSON = true;
```
Comment (andreasplesch) [inline src/nodes/Networking/Inline.js]: make default isJSON = false for backwards compatibililty

File: src/nodes/Networking/Inline.js [hunk 5]
```diff
@@ -159,9 +159,23 @@ x3dom.registerNodeType(
...
                 var xhr = new window.XMLHttpRequest();
-                if (xhr.overrideMimeType)
-                    xhr.overrideMimeType('text/xml');   //application/xhtml+xml
+                if (this._vf.url.length && this._vf.url[0].length) {
+			if (this._vf.url[0].toLowerCase().endsWith(".x3d")) {
+				isJSON = false;
+				if (xhr.overrideMimeType)
+				    xhr.overrideMimeType('text/xml');   //application/xhtml+xml
+			} else if (this._vf.url[0].toLowerCase().endsWith(".json")) {
+				if (xhr.overrideMimeType)
```
Comment (andreasplesch) [inline src/nodes/Networking/Inline.js]: isJSON -= true;

File: src/nodes/Networking/Inline.js (hunk 3 above)
Comment (andreasplesch) [reply src/nodes/Networking/Inline.js]: for now, could just assume x3d as it was the existing behaviour

**Comment by coderextreme** [comment]
I issued another pull request patching the isJSON, and hopefully restoring data urls,  please test with inline data urls and x3d.  Thanks!

John

Sent from Mail for Windows 10

**Pull request merged**